[TECH] [Pix Admin] Isoler la logique des locales des fronts dans le service "locale" (PIX-18810)

lego-technix opened the pull request:
## 🔆 Problème

Dans Pix Admin, la logique de gestion des locales et des langues est éclatée dans plusieurs fichiers.

## ⛱️ Proposition

Isoler la logique autour du service `locale` sans aucune modification du comportement actuel :
- Copier le _localeService_ `services/locale.js` de Pix App cf. #12896
- Configurer `SUPPORTED_LOCALES` : `fr`, `en`
- Remplacer toutes les récupérations de `locale`/`lang` par l’utilisation de ce _localeService_

## 🌊 Remarques

Cette PR contient des adaptations préalables du `services/locale.js` de Pix App.

## 🏄 Pour tester

* Se connecter par SSO
* Se connecter sans SSO
* Aller sur la fiche d’un utilisateur, vérifier que la _langue_ et la _locale_ affichées sont exactes
* Aller sur la fiche d’un utilisateur, modifier la _langue_ et la _locale_ et vérifier que les modifications sont bien prises en compte
* Effectuer un test d’archivage en masse (pour valider que la modification sur le `requestManagerService` est bien fonctionnelle) :

   `Administration`→`Déploiement`→`Archivage des organisations en masse`→`Importer un fichier CSV`

   Exemple de fichier CSV à utiliser : 
   
   ```csv
   ID de l'organisation
   1000
   1001
   ```

lego-technix committed:
refactor(admin): replace this.intl.setLocale by localeService.detectBestLocale

lego-technix committed:
refactor(admin): replace this.intl.primaryLocale by localeService.currentLocale

lego-technix committed:
refactor(admin): replace custom Accept Language Header by localeService.acceptLanguageHeader

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr12933.review.pix.fr)
  * [App (.org)](https://app-pr12933.review.pix.org)
  * [Orga (.fr)](https://orga-pr12933.review.pix.fr)
  * [Orga (.org)](https://orga-pr12933.review.pix.org)
  * [Certif (.fr)](https://certif-pr12933.review.pix.fr)
  * [Certif (.org)](https://certif-pr12933.review.pix.org)
  * [Junior](https://junior-pr12933.review.pix.fr)
  * [Admin](https://admin-pr12933.review.pix.fr)
  * [API](https://api-pr12933.review.pix.fr/api/)
  * [API MaDDo](https://pix-api-maddo-review-pr12933.osc-fr1.scalingo.io/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr12933.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr12933/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr12933/environment)
  * [scalingo api-maddo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr12933/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr12933/environment)

bpetetot commented on admin/tests/unit/services/request-manager-test.js:
```diff
@@ -21,8 +21,8 @@ module('Unit | Service | request-manager', function (hooks) {
 
-    intlService = this.owner.lookup('service:intl');
-    sinon.stub(intlService, 'primaryLocale').value('fr');
+    localeService = this.owner.lookup('service:locale');
+    sinon.stub(localeService, 'currentLocale').value('fr');
   });
```
Dans ce test unitaire, c'est `acceptLanguageHeader` que l'on souhaite mocker. Et on a plus besoin du `currentDomainService`

```suggestion
    localeService = this.owner.lookup('service:locale');
    sinon.stub(localeService, 'acceptLanguageHeader').value('fr');
  });
```

bpetetot replied:
Le commentaire n'a pas été pris en compte ;)

lego-technix replied:
Ah oui, zut j’avais bien vu le point _« plus besoin du currentDomainService »_ mais le point _« c'est acceptLanguageHeader que l'on souhaite mocker »_ m’avais échappé. 

C'est modifié cette fois.

bpetetot commented on admin/app/services/request-manager-handlers/locale-handler.js:
Le `currentDomain` service peut être supprimé de cette classe.

bpetetot commented on admin/tests/unit/services/locale-test.js:
```diff
@@ -0,0 +1,274 @@
+import sinon from 'sinon';
+
+const { DEFAULT_LOCALE } = ENV.APP;
+ENV.APP.SUPPORTED_LOCALES = ['en', 'es', 'fr', 'fr-BE', 'fr-FR', 'nl-BE', 'nl'];
```
Le problème de mocker le `SUPPORTED_LOCALES` globalement c'est qu'il sera modifé globalement pour tous les tests suivants ce qui peut entrainer des effets de bord sur les autres tests.

Il est préférable de le stubber dans le `beforeEach`, car il sera `restore` pour les autres tests : 
```js
hooks.beforeEach(function () {
  sinon.stub(ENV.APP, 'SUPPORTED_LOCALES').value(['en', 'es', 'fr', 'fr-BE', 'fr-FR', 'nl-BE', 'nl'])
  // ...
});
```

lego-technix replied:
Oui, très juste. Désolé 😕

lego-technix replied:
C'est corrigé. Désolé 😕

bpetetot commented on admin/app/routes/application.js:
Les services `intl` et `currentDomain` peuvent être supprimés de ce fichier

bpetetot reviewed (commented):
Dans PixAdmin, il y a le `ReferencesService` (`admin/app/services/references.js`) qui contient des langues et locales. À voir comment on pourrait le fusionner avec le locale service ou bien déclarer certaines de ces variables dans le `ENV` ?

bpetetot commented:
Quelques remarques supplémentaires :

Dans `admin/tests/integration/components/users/user-overview-test.gjs`, ce test n'a pas besoin d'itérer sur toutes les langues, seul un couple `locale`/`lang` peut être testé:
https://github.com/1024pix/pix/blob/507a8a3a331192c14d257b7f4ddd8afbe44fbaf8/admin/tests/integration/components/users/user-overview-test.gjs#L130-L158

Dans `admin/app/components/smart-random-simulator/smart-random-params.gjs`, on redéclare des locales ici:
https://github.com/1024pix/pix/blob/507a8a3a331192c14d257b7f4ddd8afbe44fbaf8/admin/app/components/smart-random-simulator/smart-random-params.gjs#L19-L24

J'ai l'impression que l'on devrait déclarer 3 listes de locales différentes dans toute les applications:
- Les locales d'affichages supportées de l'application courante
- Les locales d'affichages supportées par PixApp. Nécessaires d'être connues dans les autres apps pour certains use cases (ex: dans PixAdmin choisir éditer locale de l'utilisateur, envoyer un email à un membre d'orga dans une langue...)
- Les locales supportées par le contenu (`fr-fr`, `fr`, `en`, `es`, `nl`...)

lego-technix committed:
refactor(mon-pix): make localeService more generic wrt metricsService

lego-technix committed:
test(admin): remove useless test loop, only test one 1 locale and 1 lang

lego-technix committed:
feat(mon-pix): add pixChallengeLocales getter to localeService

This function is not currently used in Pix App but is used by other front apps.

lego-technix committed:
refactor(admin): replace SmartRandomParams.acceptedLocales by localeService.challengeLocaleOptions

lego-technix committed:
feat(mon-pix): add pixLocales getter to localeService

lego-technix committed:
feat(mon-pix): add pixLanguages getter to localeService

lego-technix committed:
refactor(mon-pix): rename so that the "available" notion for locales and languages is replaced by more meaningful notions

lego-technix committed:
refactor(mon-pix): refactor pixLanguages

lego-technix committed:
feat(admin): add new locale service copied from mon-pix

lego-technix committed:
refactor(admin): replace references.availableLocales by localeService.pixLocales

lego-technix committed:
refactor(admin): replace references.availableLanguages by localeService.pixLanguages

lego-technix committed:
refactor(admin): remove referencesService now useless

lego-technix commented:
> Quelques remarques supplémentaires :
> Dans `admin/tests/integration/components/users/user-overview-test.gjs`, ce test n'a pas besoin d'itérer sur toutes les 
> […]
> Dans `admin/app/components/smart-random-simulator/smart-random-params.gjs`, on redéclare des locales ici:
> […]
> J'ai l'impression que l'on devrait déclarer 3 listes de locales différentes dans toute les applications:
> […]
>

Tout ces points ont été pris en compte.

bpetetot commented on mon-pix/app/services/locale.js:
```diff
@@ -44,6 +44,14 @@ export default class LocaleService extends Service {
 
+  /**
+   * Returns all languages available in the Pix platform
+   * @deprecated use locales instead whenever possible.
```
```suggestion
   * @deprecated use `pixLocales` instead whenever possible.
```

lego-technix replied:
`@deprecated use pixLocales` mis en place.

bpetetot reviewed (approved):
✅ Revu et testé 👍

lego-technix committed:
docs(mon-pix): improve code comments wrt ORIGINAL file and COPY files

lego-technix committed:
docs(admin): improve code comments wrt ORIGINAL file and COPY files

EmmanuelleBonnemay reviewed (approved):
Lu et testé avec succès (local et RA)

pix-service commented:
:tada: This PR is included in version 5.167.0 :tada:

The release is available on [GitHub release](https://github.com/1024pix/pix/releases/tag/v5.167.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
